### ISSUE TITLE
predict() function for logistic regression classifier is broken

### ISSUE THREAD

**Issue by afriesen** [Itseez/opencv#5413]
There are two issues with `float LogisticRegressionImpl::predict(InputArray samples, OutputArray results, int) const` (in lr.cpp).

1) This function always returns 0 (there's a TODO at the bottom of the function that is probably about fixing this).

2) This function always attempts to write to `OutputArray results` even if `results = noArray()`, which causes a segfault. 

These issues both arise when calling `StatModel::calcError( ... )` on a logistic regression classifier, since this function specifies `OutputArray results = noArray()` and expects the returned value to be meaningful.

**Comment by StevenPuttemans** [comment]
Probably the fastest way to get this fixed is by providing a bugfix yourself and trying to finish the TODO. The second problem can be solved by checking the content of results. Are you up for it?

**Comment by afriesen** [comment]
Yes, I can take a stab at this. 

A related question in lr.cpp: I see that in `compute_batch_gradient`, the for loop ranges over `params.num_iters` whereas in `compute_mini_batch_gradient`, the for loop ranges over `term_crit.maxCount`. By default, these have the same value, but they can have different values if specified. Is there a reason that these use different parameters?

Similarly, in my understanding of mini batch gradient descent, one still wants to iterate until convergence or the iteration limit is reached; however, in `compute_mini_batch_gradient` there's an if-statement at the end (`if(j+size_b>_data.rows)`) that will terminate this after a single iteration. Does this exist for a reason or is it a bug?

**Comment by afriesen** [comment]
Further, the convergence criteria and the learning rate are typically separate parameters, but the LR code here uses the same value for both (by default). Is this the intended behavior?

Actually, convergence doesn't seem to even be checked here, even though `TermCriteria` is specified as `COUNT + EPS`.

**Comment by StevenPuttemans** [comment]
@afriesen I am sorry but my knowledge of the convergence algorithms is rather limited. I assume that if you provide a PR with suggested adaptations that you think are needed, a more knowledged OpenCV developer will get assigned. Then you can fire away all your questions there!

**Comment by Dikay900** [comment]
can be closed, PR was merged
